Document and test allowed identifiers for variables (context references) and `.` syntax

The [docs](https://json-e.js.org/expressions.html#context-references) don't give any indication as to what's valid for the dot syntax in bar identifiers.

For example, in `x.a`, what is valid for the `a`?

In JSON Path, we've decided to limit valid identifiers to `^[a-zA-Z_][a-zA-Z0-9_]*$`.  That is, start with alpha or underscore, and the rest must be alphanumeric or underscore.

Exploring in the playground reveals this error:

```
Template: {$eval: 'b.a@4'}
Context:  {b: {'a@4': 5}}

SyntaxError: Unexpected input for 'b.a@4' at '@4'
```

What's valid for this syntax is somewhat discoverable, but it should probably be documented.

### From the tests:

```
title: invalid context
context: {'a b c': 1}
template: {}
error: 'TemplateError: top level keys of context must follow /[a-zA-Z_][a-zA-Z0-9_]*/'
panic: true
```

There aren't any tests for nested properties.

--- a/docs/src/expressions.md
+++ b/docs/src/expressions.md
@@ -47,6 +47,11 @@ context: {x: 'quick', z: 'sort'}
 result: ['quick', 'sort', 'quicksort']
 ```
 
+Valid identifiers for context references follow the requirements of many
+programming languages: identifiers may only contain letters, underscores, and
+numbers, but may not start with a number. This does imply that the context
+object may only have keys that meet these requirements.
+
 ## Arithmetic Operations
 
 The usual arithmetic operators are all defined, with typical associativity and
@@ -128,8 +133,8 @@ does not have `prop`, while `obj['prop']` will evaluate to `null`.
 
 ```yaml,json-e
 template: {$eval: 'v.a + v["b"]'}
-context: {v: {a: 'apple', b: 'bananna', c: 'carrot'}}
-result: 'applebananna'
+context: {v: {a: 'apple', b: 'banana', c: 'carrot'}}
+result: 'applebanana'
 ```
 
 Note that the object can be a literal expression:
@@ -139,6 +144,13 @@ template: {$eval: '{ENOMEM:"Out of memory", ENOCPU:"Out of CPUs"}[msgid]'}
 context: {msgid: ENOMEM}
 result: 'Out of memory'
 ```
+
+When using the dot-syntax, e.g. `.prop`, identifiers may only contain letters,
+underscores, and numbers, but may not start with a number (just like for context
+references). While the context object may only have keys that meet these
+requirements, nested objects may have keys in any format. Keys that are not in
+the identifier format must be accessed using the bracket-name syntax, e.g.
+`['my-prop']`.
 
 ## Indexing and Slicing
 

--- a/internal/interpreter/util.go
+++ b/internal/interpreter/util.go
@@ -273,7 +273,7 @@ var contextVariablePattern = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*$`)
 func IsValidContext(context map[string]interface{}) error {
 	for k, v := range context {
 		if !contextVariablePattern.MatchString(k) {
-			return fmt.Errorf("context variable '%s' doesn't match pattern: %s", k, contextVariablePattern.String())
+			return fmt.Errorf("top level keys of context must follow %s", contextVariablePattern.String())
 		}
 		if err := IsValidData(v); err != nil {
 			return err

--- a/internal/jsone.go
+++ b/internal/jsone.go
@@ -19,7 +19,7 @@ import (
 func Render(template interface{}, context map[string]interface{}) (interface{}, error) {
 	// Validate input
 	if err := i.IsValidContext(context); err != nil {
-		panic(err)
+		return nil, err
 	}
 
 	// Inherit functions from builtins

--- a/internal/jsone_test.go
+++ b/internal/jsone_test.go
@@ -19,18 +19,10 @@ type testCase struct {
 	Template interface{}            `json:"template"`
 	Result   interface{}            `json:"result"`
 	Error    interface{}            `json:"error"` // bool, string or nil
-	Panic    interface{}            `json:"panic"` // bool, true if panic is expected
 }
 
 func (c *testCase) Test(t *testing.T) {
 	require.Empty(t, c.Section, "sections aren't test cases")
-
-	if c.Panic != nil {
-		require.Panics(t, func() {
-			Render(c.Template, c.Context) // nolint:errcheck
-		})
-		return
-	}
 
 	// Set a fixed 'now', if one isn't specified
 	context := make(map[string]interface{})

--- a/rs/src/render.rs
+++ b/rs/src/render.rs
@@ -5,6 +5,14 @@ use crate::interpreter::{self, Context};
 use crate::op_props::{parse_by, parse_each};
 use crate::value::{Object, Value};
 use anyhow::{bail, Result};
+use nom::{
+    branch::alt,
+    bytes::complete::tag,
+    character::complete::{alpha1, alphanumeric1},
+    combinator::{all_consuming, recognize},
+    multi::many0,
+    sequence::pair,
+};
 use serde_json::Value as SerdeValue;
 use std::borrow::Cow;
 use std::convert::TryInto;
@@ -750,23 +758,18 @@ fn sort_operator_without_by(
 
 /// Recognize identifier strings for $let
 pub(crate) fn is_identifier(identifier: &str) -> bool {
-    let mut chars = identifier.chars();
+    fn parser(input: &str) -> nom::IResult<&str, &str> {
+        all_consuming(recognize(pair(
+            alt((alpha1, tag("_"))),
+            many0(alt((alphanumeric1, tag("_")))),
+        )))(input)
+    }
 
-    if let Some(c) = chars.next() {
-        if !c.is_ascii_alphabetic() {
-            return false;
-        }
+    if let Ok((remaining, _)) = parser(identifier) {
+        remaining.is_empty()
     } else {
-        return false;
+        false
     }
-
-    for c in chars {
-        if !c.is_ascii_alphanumeric() && c != '_' {
-            return false;
-        }
-    }
-
-    return true;
 }
 
 #[cfg(test)]

--- a/specification.yml
+++ b/specification.yml
@@ -4031,6 +4031,11 @@ context: {key: 'null'}
 template: {$eval: 'key'}
 result: 'null'
 ---
+title: 'property that is also a function name'
+context: {min: abc}
+template: {$eval: 'min'}
+result: abc
+---
 title: 'missing property'
 context: {key: {a: 1}}
 template: {$eval: 'key.b'}
@@ -4076,6 +4081,81 @@ context: {key: abc, values: {abc: 2, def: 3}}
 template: {$eval: 'values[key]'}
 result: 2
 ---
+title: 'property starting with underscore'
+context: {_key: abc}
+template: {$eval: '_key'}
+result: abc
+---
+title: 'property containing underscore'
+context: {my_key: abc}
+template: {$eval: 'my_key'}
+result: abc
+---
+title: 'property containing hyphen'
+context: {my-key: abc}
+template: {$eval: 'my-key'}
+error: 'TemplateError: top level keys of context must follow /[a-zA-Z_][a-zA-Z0-9_]*/'
+---
+title: 'property starting with capital'
+context: {Key: abc}
+template: {$eval: 'Key'}
+result: abc
+---
+title: 'property containing capital'
+context: {kEy: abc}
+template: {$eval: 'kEy'}
+result: abc
+---
+title: 'case sensitivity of keys'
+context: {key: abc}
+template: {$eval: 'Key'}
+error: 'InterpreterError: unknown context value Key'
+---
+title: 'property starting with number'
+context: {'2fast': abc}
+template: {$eval: '2fast'}
+error: 'TemplateError: top level keys of context must follow /[a-zA-Z_][a-zA-Z0-9_]*/'
+---
+title: 'property containing number'
+context: {furious5: abc}
+template: {$eval: 'furious5'}
+result: abc
+---
+title: 'nested property starting with underscore'
+context: {key: {_foo: abc}}
+template: {$eval: 'key._foo'}
+result: abc
+---
+title: 'nested property containing underscore'
+context: {key: {foo_bar: abc}}
+template: {$eval: 'key.foo_bar'}
+result: abc
+---
+title: 'nested property containing hyphen'
+context: {key: {my-foo: abc}}
+template: {$eval: 'key.my-foo'}
+error: 'InterpreterError: object has no property "my"'
+---
+title: 'nested property containing hyphen using brackets'
+context: {key: {my-foo: abc}}
+template: {$eval: "key['my-foo']"}
+result: abc
+---
+title: 'nested property starting with capital'
+context: {key: {Foo: abc}}
+template: {$eval: 'key.Foo'}
+result: abc
+---
+title: 'nested property containing capital'
+context: {key: {fOo: abc}}
+template: {$eval: 'key.fOo'}
+result: abc
+---
+title: 'case sensitivity of nested keys'
+context: {key: {foo: abc}}
+template: {$eval: 'key.Foo'}
+error: 'InterpreterError: object has no property "Foo"'
+---
 title: 'property access with number'
 context:
   obj: {'13': 15}
@@ -4104,6 +4184,11 @@ template: {$eval: 'key.valueOf'}
 error: 'InterpreterError: infix: . expects objects'
 ---
 title: 'non-identifier after dot'
+context: {key: {}}
+template: {$eval: 'key.{'}
+error: 'SyntaxError: Found: { token, expected one of: identifier'
+---
+title: 'non-identifier after dot (2)'
 context: {key: {}}
 template: {$eval: 'key.{'}
 error: 'SyntaxError: Found: { token, expected one of: identifier'

--- a/specification.yml
+++ b/specification.yml
@@ -126,7 +126,6 @@ title: invalid context
 context: {'a b c': 1}
 template: {}
 error: 'TemplateError: top level keys of context must follow /[a-zA-Z_][a-zA-Z0-9_]*/'
-panic: true
 ################################################################################
 ---
 section:  $if operator

